include/nuttx/msgq.h

```diff
@@ -0,0 +1,315 @@
+ *
+ * Input Parameters:
+ *   buffer   - Ring buffer of Message Queue
+ *   msg_size - Message size (in bytes).
+ *   max_msgs - Maximum number of messages that can be queued.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+#  define NXMSGQ_INITIALIZER(buffer, msg_size, max_msgs) \
+    { \
+      CIRCBUF_INITIALIZER(buffer, msg_size * max_msgs), \
+      msg_size, \
+      NXSEM_INITIALIZER(0, 0), \
+      NXSEM_INITIALIZER(0, 0), \
+      false, \
+      SP_UNLOCKED, \
+    }
+#else
+#  define NXMSGQ_INITIALIZER(buffer, msg_size, max_msgs) \
+    { \
+      CIRCBUF_INITIALIZER(buffer, msg_size * max_msgs), \
+      msg_size, \
+      NXSEM_INITIALIZER(0, 0), \
+      NXSEM_INITIALIZER(0, 0), \
+      false, \
+    }
+#endif
+
+/****************************************************************************
+ * Public Type Declarations
+ ****************************************************************************/
+
+/* This structure defines a kernel message queue */
+
+typedef struct nxmsgq
+{
+  struct circbuf_s cbuf;
+  int              msg_size;
+  sem_t            txsem;
+  sem_t            rxsem;
+  bool             alloc;
+#ifdef CONFIG_SMP
+  spinlock_t       lock;
+#endif
+} nxmsgq_t;
+
+/****************************************************************************
+ * Public Function Prototypes
+ ****************************************************************************/
+
+#ifdef __cplusplus
+#define EXTERN extern "C"
+extern "C"
+{
+#else
+#define EXTERN extern
+#endif
+
+/****************************************************************************
+ * Name: nxmsgq_init
+ *
+ * Description:
+ *   Initialize a message queue with static ring buffer
+ *
+ *   This routine initializes a message queue object, prior to its first use.
+ *
+ *   The message queue's ring buffer must contain space for max_msgs
+ *   messages, each of which is msg_size bytes long. Alignment of the
+ *   message queue's ring buffer is not necessary.
+ *
+ * Input Parameters:
+ *   msgq     - Address of the message queue.
+ *   buffer   - Pointer to ring buffer that holds queued messages.
+ *   msg_size - Message size (in bytes).
+ *   max_msgs - Maximum number of messages that can be queued.
+ *
+ * Returned Value:
+ *   None.
+ *
+ ****************************************************************************/
+
+void nxmsgq_init(FAR nxmsgq_t *msgq, FAR void *buffer,
+                 size_t msg_size, uint32_t max_msgs);
+
+/****************************************************************************
+ * Name: nxmsgq_create
+ *
+ * Description:
+ *   Create a message queue.
+ *
+ *   This routine initializes a message queue object, prior to its first use,
+ *   allocating its internal ring buffer from the calling thread's resource
+ *   pool.
+ *
+ *   Memory allocated for the ring buffer can be released by calling
+ *   nxmsgq_destroy()
+ *
+ * Input Parameters:
+ *   msg_size - Message size (in bytes).
```

Comment:
Does this mean that a nxmsgq can only handle messages of a certain size ? Isn't this limitation a bit too harsh ?

Reply:
Yes, this is just aligned with other operating systems

FreeRTOS:
https://freertos.org/Documentation/02-Kernel/04-API-references/06-Queues/02-xQueueCreateStatic
ThreadX:
https://github.com/eclipse-threadx/rtos-docs/blob/main/rtos-docs/threadx/chapter4.md#tx_queue_create